### PR TITLE
[C++] use wrapper over 'memset' that checks for null ptr

### PR DESCRIPTION
**Pull request by kevinushey**
One part of https://github.com/wesm/feather/issues/170.

### CODE REVIEW

File: R/src/feather-write.cpp [hunk 1]
```diff
@@ -106,12 +106,12 @@ PrimitiveArray factorCodesToPrimitiveArray(SEXP x) {
 
   auto values_buffer = std::make_shared<OwnedMutableBuffer>();
   stopOnFailure(values_buffer->Resize(n * sizeof(int32_t)));
-  memset(values_buffer->mutable_data(), 0, n);
+  util::fill_buffer(values_buffer->mutable_data(), 0, n);
   auto values = reinterpret_cast<int32_t*>(values_buffer->mutable_data());
 
   for (int i = 0; i < n; ++i) {
-    // Results of NA values don't matter as the value slots are undefined
```
Comment (kevinushey) [inline R/src/feather-write.cpp]: Does this comment effectively mean it doesn't matter what an NA value is serialized as? (I guess they're tracked in some alternate mechanism?)

Comment (wesm) [inline R/src/feather-write.cpp]: There's a validity bitmap (1 for valid, 0 for NA). This should be used on reconstructing the factor codes (setting whatever is the NA code in the slot)

File: R/src/feather-types.cpp [hunk 2]
```diff
@@ -213,8 +213,9 @@ SEXP rescaleFromInt64(const PrimitiveArray* pArray, double scale = 1) {
 template <typename T>
 static void write_factor_codes(const uint8_t* values, int length, int* out) {
   auto codes = reinterpret_cast<const T*>(values);
+  T maximum = std::numeric_limits<T>::min();
   for (int i = 0; i < length; ++i) {
-    out[i] = codes[i] + 1;
+    out[i] = codes[i] + (codes[i] != maximum);
```
Comment (wesm) [inline R/src/feather-types.cpp]: This should perhaps use the validity bitmap?

Comment (wesm) [inline R/src/feather-types.cpp]: I need to look a little closer at this code and will get back to you

Comment (kevinushey) [inline R/src/feather-types.cpp]: Thanks; like you mentioned there's probably a solution that uses the validity bitmap but wasn't entirely sure.

**Comment by wesm** [comment]
I'm going to make a pull request soon (with a couple unit tests hopefully) that supercedes this one

**Pull request closed without merging**